### PR TITLE
VisualElement refactor

### PR DESCRIPTION
**Pull request by InventingWithMonster**
This PR removes the `stylefire` dependency and introduces the `VisualElement` class. A `VisualElement` is a generic adapter between Framer Motion and different renderers. Currently we only have HTML and SVG renderers but this will be expanded to others. By dealing with VisualElement the hope is that we can extend functionality to other renders simply by writing new VisualElements. 

This PR isn't fully generic but provides enough of a refactor that we'll better identify where further progress can be made when we make another renderer.

By removing `stylefire` as a dependency and moving the direct renderer to within Framer Motion, we're also making it possible to refactor the layout animation API to be additive. This means we can first animate and/or correct the layout, then additionally animate the transform if this has been defined by the user.

It also reduces generated bundle size by ~1.5kb

### CODE REVIEW

File: dev/examples/visualElement.tsx [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+        ? { x, y: 10, backgroundColor: "red" }
+        : { x: -100, y: 50, backgroundColor: "green" }
+
+    return (
+        <motion.div
+            onClick={() => setIs(!is)}
+            style={{ ...style, ...motionStyle }}
+        />
+    )
+}
```
Comment (hemlok) [inline dev/examples/visualElement.tsx:26]: Should this example animate?

File: src/animation/use-animated-state.ts [hunk 2]
```diff
@@ -1,67 +1,56 @@
...
+    }
+
+    readNativeValue(key: string) {
+        return this.latestState[key] || 0
+    }
+
+    render() {
+        console.log("render", this.latest.foo)
+    }
+}
```
Comment (hemlok) [inline src/animation/use-animated-state.ts:25]: Is this WIP code? Can it be committed safely?

File: src/components/AnimatePresence/use-presence.ts [hunk 3]
```diff
@@ -34,11 +34,20 @@ type NotPresent = [false, SafeToRemove]
...
     return !isPresent && onExitComplete ? [false, onExitComplete] : [true]
 }
+
+/**
+ * @public
+ */
+export function useIsPresent() {
+    const context = useContext(PresenceContext)
+    return context === null ? true : context.isPresent
+}
```
Comment (hemlok) [inline src/components/AnimatePresence/use-presence.ts:53]: 👍

File: src/motion/__tests__/variant.test.tsx [hunk 4]
```diff
@@ -39,474 +39,474 @@ describe("animate prop as variant", () => {
...
-        }
-
-        const { rerender } = render(<Component length={1} />)
-        rerender(<Component length={1} />)
-        rerender(<Component length={2} />)
-        rerender(<Component length={2} />)
-
-        expect(x.get()).toBe(100)
-    })
+    // test("fires onAnimationStart when animation begins", async () => {
```
Comment (hemlok) [inline src/motion/__tests__/variant.test.tsx:42]: 😂 What's the deal with these tests? Not valuable anymore? Or just not compatible with Visual Element yet and need to be updated down the road?

File: dev/examples/visualElement.tsx (hunk 1 above)
Comment (InventingWithMonster) [reply dev/examples/visualElement.tsx:26]: No, actually I'll delete it as I was just ensuring styles get applied/removed

File: src/animation/use-animated-state.ts (hunk 2 above)
Comment (InventingWithMonster) [reply src/animation/use-animated-state.ts:25]: Ah yeah good catch on the render. This is basically just some hack to reimplement useAnimatedState using the new internal APIs. This isn't a public API, but some Framer users were told about it and now use it in their projects.

File: src/motion/__tests__/variant.test.tsx (hunk 4 above)
Comment (InventingWithMonster) [reply src/motion/__tests__/variant.test.tsx:42]: Ha no I forgot to uncomment them. I was testing just the one test at the time.

File: src/animation/use-animated-state.ts (hunk 2 above)
Comment (hemlok) [reply src/animation/use-animated-state.ts]: Cool 👍

File: src/render/dom/HTMLVisualElement.ts [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+
+    /**
+     * Render the Element by rebuilding and applying the latest styles and vars.
+     */
+    render() {
+        // Rebuild the latest animated values into style and vars caches.
+        this.build()
+
+        // Directly assign style into the Element's style prop. In tests Object.assign is the
+        // fastest way to assign styles.
```
Comment (hemlok) [inline src/render/dom/HTMLVisualElement.ts:180]: Good to know 👀

**Comment by hemlok** [comment]
Code looks good 🧙  Going to run through the examples now.

**Comment by hemlok** [comment]
Just ran through all the examples. Most everything is unchanged from before 🎉 

There are 3 examples that seem to be not functioning the same as on `feature/magic-move`: `animateSVGPath`, `heightAutoFromDisplayNone`, `viewportScroll`.

**Pull request merged**